File: ripperdoc/wetware/imprint/scripts/inject_hook.py
#!/usr/bin/env python3
"""Context-injection hook — inject an imprint template with host-specific tooling info.

Wired inline in each host's manifest:
  .claude-plugin/plugin.json   UserPromptSubmit + SessionStart   — Claude
  .codex-plugin/plugin.json    UserPromptSubmit + SessionStart   — Codex
  gemini-extension.json        BeforeAgent      + SessionStart   — Gemini

Args (positional, backwards-compatible):
  argv[1] — prompt template path (relative to plugin root) OR `--file <name>`
            where <name> is a short alias (`imprint`, `visuals`) that maps to
            `<NAME>.md` in the plugin root
  argv[2] — host-tools file path (relative to plugin root); contents replace
            `{{host_tools}}` in the template. Optional — templates without the
            placeholder (e.g. VISUALS.md) ignore it.

Examples:
  inject_hook.py IMPRINT.md hosts/host_claude.md         # legacy positional
  inject_hook.py --file imprint hosts/host_claude.md     # equivalent via alias
  inject_hook.py --file visuals                          # static template, no host-tools

The hook reads `hook_event_name` from the JSON payload on stdin and echoes it
back in the envelope, so the same script works for any context-injection event
(SessionStart, UserPromptSubmit, BeforeAgent, …) without code changes — each
manifest is the only place that knows which event to wire.

  --event <name>  fallback event name when stdin omits `hook_event_name`
                  (default `UserPromptSubmit`). stdin always wins when it
                  carries a non-empty `hook_event_name`. Lets each manifest
                  declare the right label for its wiring (e.g. SessionStart,
                  BeforeAgent) instead of silently defaulting.

Always exits 0 — missing or empty files are silent no-ops; the host never
blocks on this hook. Missing host-tools file substitutes to empty.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

PLUGIN_ROOT = Path(__file__).resolve().parents[1]
DEFAULT_EVENT = "UserPromptSubmit"


def extract_event_arg(argv: list[str]) -> tuple[str | None, list[str]]:
    """Pull `--event <name>` out of argv; return (event_or_None, remaining_argv).

    Extracted before template resolution so positional/`--file` parsing sees
    argv with the flag already removed. Only the first `--event` is honored.
    """
    out: list[str] = []
    event: str | None = None
    i = 0
    while i < len(argv):
        if argv[i] == "--event" and event is None:
            if i + 1 < len(argv):
                event = argv[i + 1]
                i += 2
                continue
            i += 1
            continue
        out.append(argv[i])
        i += 1
    return event, out


def resolve_template_arg(argv: list[str]) -> tuple[str | None, list[str]]:
    """Resolve argv[1] into a template path; return (path, remaining_argv).

    Supports two forms:
      --file <name>   → <NAME>.md (alias for short-name invocation)
      <path>          → used verbatim (legacy positional)
    """
    if not argv:
        return None, []
    if argv[0] == "--file":
        if len(argv) < 2:
            return None, []
        return f"{argv[1].upper()}.md", argv[2:]
    return argv[0], argv[1:]


def main() -> int:
    cli_event, argv = extract_event_arg(sys.argv[1:])
    template_rel, rest = resolve_template_arg(argv)
    if template_rel is None:
        return 0

    prompt_path = PLUGIN_ROOT / template_rel
    if not prompt_path.is_file():
        return 0
    template = prompt_path.read_text(encoding="utf-8")

    host_tools = ""
    if rest:
        host_path = PLUGIN_ROOT / rest[0]
        if host_path.is_file():
            host_tools = host_path.read_text(encoding="utf-8").strip()

    text = template.replace("{{host_tools}}", host_tools).strip()
    if not text:
        return 0

    default_event = cli_event or DEFAULT_EVENT
    event_name = default_event
    try:
        payload = json.loads(sys.stdin.read() or "{}")
        if isinstance(payload, dict):
            name = payload.get("hook_event_name")
            if isinstance(name, str) and name:
                event_name = name
    except (json.JSONDecodeError, OSError):
        pass

    print(
        json.dumps(
            {
                "hookSpecificOutput": {
                    "hookEventName": event_name,
                    "additionalContext": text,
                }
            }
        )
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: ripperdoc/wetware/imprint/scripts/inject-hook.ts
#!/usr/bin/env -S preemdeck-bun
/**
 * inject-hook.ts — imprint-template context injector.
 *
 * Resolves a template (positional path, or `--file <name>` -> <NAME>.md), reads it
 * from the plugin root, substitutes the optional host-tools file's contents for
 * `{{host_tools}}`, strips, and injects via lib/hook.ts. Missing/empty files are a
 * silent `{}` no-op; a missing host-tools file substitutes empty. Default event
 * UserPromptSubmit; `--event <name>` (first only) is the fallback; stdin wins.
 *
 * Path note: args resolve as `PLUGIN_ROOT / arg` with an "absolute arg wins"
 * rule — Node's `resolve()` honors absolute temp paths verbatim. PLUGIN_ROOT =
 * <script-dir>/.. (scripts/ -> imprint/).
 */

import { readFile, stat } from "node:fs/promises"
import { dirname, resolve } from "node:path"
import { exists } from "../../../../lib/fs.ts"
import { runInjectionHook } from "../../../../lib/inject.ts"

const PLUGIN_ROOT = dirname(import.meta.dir)
// (The UserPromptSubmit default lives in lib/inject.ts; the --event flag overrides it.)

/**
 * Pull `--event <name>` out of argv; return [event_or_null, remaining_argv].
 * Only the first `--event` is honored.
 */
export const extractEventArg = (argv: string[]): [string | null, string[]] => {
  const out: string[] = []
  let event: string | null = null
  let i = 0
  while (i < argv.length) {
    if (argv[i] === "--event" && event === null) {
      if (i + 1 < argv.length) {
        event = argv[i + 1] as string
        i += 2
        continue
      }
      i += 1
      continue
    }
    out.push(argv[i] as string)
    i += 1
  }
  return [event, out]
}

/**
 * Resolve argv[0] into a template path; return [path_or_null, remaining_argv].
 *   --file <name>  -> <NAME>.md (uppercased)
 *   <path>         -> used verbatim
 */
export const resolveTemplateArg = (argv: string[]): [string | null, string[]] => {
  if (argv.length === 0) return [null, []]
  if (argv[0] === "--file") {
    if (argv.length < 2) return [null, []]
    return [`${(argv[1] as string).toUpperCase()}.md`, argv.slice(2)]
  }
  return [argv[0] as string, argv.slice(1)]
}

const isFile = async (path: string): Promise<boolean> => {
  return (await exists(path)) && (await stat(path)).isFile()
}

/**
 * Build the injected text from argv (the script's tail). Returns the stripped
 * text, or null for any no-op (no template arg, missing/empty template, empty
 * after substitution+strip). `pluginRoot` defaults to the real plugin root.
 */
export const renderTemplate = async (argv: string[], pluginRoot: string = PLUGIN_ROOT): Promise<string | null> => {
  const [templateRel, rest] = resolveTemplateArg(argv)
  if (templateRel === null) return null

  const promptPath = resolve(pluginRoot, templateRel)
  if (!(await isFile(promptPath))) return null
  const template = await readFile(promptPath, "utf8")

  let hostTools = ""
  if (rest.length > 0) {
    const hostPath = resolve(pluginRoot, rest[0] as string)
    if (await isFile(hostPath)) {
      hostTools = (await readFile(hostPath, "utf8")).trim()
    }
  }

  const text = template.replaceAll("{{host_tools}}", hostTools).trim()
  return text || null
}

if (import.meta.main) {
  const [cliEvent, argv] = extractEventArg(Bun.argv.slice(2))
  const text = await renderTemplate(argv)
  await runInjectionHook({
    event: cliEvent ?? undefined,
    render: () => text,
  })
  process.exit(0)
}
